pageserver/src/basebackup.rs

```diff
@@ -154,9 +154,17 @@ impl<'a> Basebackup<'a> {
             let img = self
                 .timeline
                 .get_slru_page_at_lsn(slru, segno, blknum, self.lsn)?;
-            ensure!(img.len() == pg_constants::BLCKSZ as usize);
 
-            slru_buf.extend_from_slice(&img);
+            if slru == SlruKind::Clog {
+                ensure!(
+                    img.len() == pg_constants::BLCKSZ as usize
+                        || img.len() == pg_constants::BLCKSZ as usize + 8
```

Comment:
I don't think it makes any difference from an I/O and caching point of view, given our current storage system. The key+value pairs are not stored in an aligned fashion in the repository anyway. And compression, once we add it, will break it anyway.

It would not fit nicely into the materialized page cache though. But we don't use it for clog pages currently, and if we want to do that in the future, we can leave the timestamp out of the cache, for example.

Reply:
> It feels like page granularity precision is too coarse. with one timestamp for 32K transactions. Should we just track XID->TS separately? I imagine can be pretty efficiently compressible and won't take much RAM

We could track it separately, but that adds some overhead to processing commit records. Timestamp->LSN mapping is a pretty rare operation, I think @knizhnik is right that we should optimize for keeping the overhead on WAL processing minimal, even if it makes timestamp->LSN mapping more expensive.

Currently, we don't support PostgreSQL commit timestamp tracking (`track_commit_timestamp` setting). One approach would be to implement support for that, i.e. store pg_commit_ts also in the repository, and then use that to find the timestamp. We'd still need to do the binary search with LSN though. pg_commit_ts maps from XID to timestamp, and we need to map timestamp to LSN.

Reply:
> @antons-antons  It feels like page granularity precision is too coarse. with one timestamp for 32K transactions. 

May be there is some misunderstanding here... Please notice that attaching latest timestamp to CLOG page doesn't mean that we are able to locate by timestamp only this latest transaction (one of 32k transactions).If it was true, then definitely this approach can not be used: XID and commit timestamps are completely unrelated and at the same CLOG page there can transactions with hours of delta between their commit time.

But we are using binary search (with LSN as key) and reconstruct CLOG page at the particular LSN.
So each time we materialize page, we get different timestamp. So we are able to locate closest timestamp of the particular transaction. So Heikki's approach is not less precise then my approach.

There is just one source of uncertainty in both cases: there is no warranty in Postgres that order of transaction commit timestamps is the same as their commit order (or order of their LSNs). I.e. it can happen that transaction T1 is committed before T2, but has larger commit timestamp. but in this case this delta between T1 and T2 commit timestamps will be very small. And from user's point of view there is no sense to distinguish state between T1 and T2 using timestamp.

> @hlinnaka Currently, we don't support PostgreSQL commit timestamp tracking 

I do not completely understand you answer. Yes, we do not support  PostgreSQL commit timestamp tracking. But we actually can do it on top of current timestmp->LSN mapping. I do not see any sense in opposite solution: somehow make Postgres commit timestamp tracking mechanism to persist its changes at page server and implement timestmp->LSN mapping using this mechanism.

Reply:
PostgreSQL commit timestamp tracking means just presence of few functions: pg_xact_commit_timestamp,...
We can implement this functions without maintaining pg_commit_ts SLRU. So we will be compatible with Postgres and do not pay extra price for it (unlike Vanilla).